File: .agents/skills/simulate-table/scripts/deal-table.ts
#!/usr/bin/env bun

import { mkdir, readdir, readFile, writeFile } from "node:fs/promises"
import { basename, dirname, join, relative, resolve } from "node:path"
import { createInterface } from "node:readline/promises"

type Json = Record<string, any>

export type Deck = {
  path: string
  relativePath: string
  title: string
  commanders: string[]
  plan: string
}

type Candidate = {
  mulligans: number
  hand: string[]
  library: string[]
}

type Seat = Deck & {
  id: string
  color: string
  entries: Json[]
  candidates: Candidate[]
  tokenSources: Record<string, string[]>
  tokens: Record<string, Json>
}

type Options = {
  apply: boolean
  bottom: string[]
  deckQueries: string[]
  format: "markdown" | "json"
  list: boolean
  mulligans: string
  out?: string
  repo: string
  seed: number
  turns: number
}

export const ROOT = resolve(import.meta.dir, "../../../..")
export const SEAT_IDS = ["p1", "p2", "p3", "p4"]
const SEAT_COLORS = ["#c45c26", "#2f6f64", "#4a5d9e", "#8a3d6b"]
const MULLIGAN_MAX = 2

const normalize = (value: string) =>
  value
    .normalize("NFKD")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, " ")
    .trim()

export const readJson = async (path: string) => JSON.parse(await readFile(path, "utf8"))

const readPrimer = async (deckPath: string) => {
  try {
    return await readFile(join(deckPath, "README.md"), "utf8")
  } catch {
    return ""
  }
}

const titleFromPrimer = (primer: string, fallback: string) => {
  const first = primer.split(/\r?\n/, 1)[0] ?? ""
  return first.match(/—\s+(.+)$/)?.[1]?.trim() || fallback
}

const planFromPrimer = (primer: string) => {
  const line = primer.split(/\r?\n/).find((entry) => entry.startsWith("**Primary plan:**"))
  return (line ?? "")
    .replace(/\[([^\]]+)\]\([^)]+\)/g, "$1")
    .replace(/\*\*/g, "")
    .replace("Primary plan:", "")
    .trim()
}

const isLibraryCard = (entry: Json) =>
  !(entry.categories ?? []).some(
    (category: string) => category === "Commander{top}" || category.includes("{noDeck}"),
  )

export const loadManifest = async (deckPath: string) => {
  const manifestPath = join(deckPath, "cards.json")
  const manifest = await readJson(manifestPath).catch(() => {
    throw new Error(`missing resolved manifest: ${manifestPath}`)
  })
  const library: string[] = []
  const commanders: string[] = []
  const entries: Json[] = []

  for (const entry of manifest.cards ?? []) {
    const { name, quantity, categories = [] } = entry
    if (typeof name !== "string" || !Number.isInteger(quantity) || quantity < 1) {
      throw new Error(`invalid manifest entry in ${manifestPath}`)
    }
    entries.push(entry)
    if (categories.includes("Commander{top}")) {
      commanders.push(...Array(quantity).fill(name))
    } else if (isLibraryCard(entry)) {
      library.push(...Array(quantity).fill(name))
    }
  }

  if (![98, 99].includes(library.length)) {
    throw new Error(`${deckPath}: expected a 98- or 99-card library, found ${library.length}`)
  }
  if (![1, 2].includes(commanders.length)) {
    throw new Error(`${deckPath}: expected one or two commanders, found ${commanders.length}`)
  }
  if (library.length + commanders.length !== 100) {
    throw new Error(`${deckPath}: expected 100 total cards`)
  }

  return { library, commanders, entries }
}

export const discoverDecks = async (repo: string) => {
  const decksPath = join(repo, "decks")
  const entries = await readdir(decksPath, { withFileTypes: true })
  const decks: Deck[] = []

  for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
    if (!entry.isDirectory()) continue
    const path = join(decksPath, entry.name)
    try {
      const primer = await readPrimer(path)
      const { commanders } = await loadManifest(path)
      decks.push({
        path,
        relativePath: relative(repo, path),
        title: titleFromPrimer(primer, entry.name),
        commanders,
        plan: planFromPrimer(primer),
      })
    } catch {
      // Unresolved deck workspaces are intentionally not selectable.
    }
  }

  return decks
}

const deckLabel = (deck: Deck) =>
  `${deck.title} — ${deck.commanders.join(" + ")} (${basename(deck.path)})`

const matchesDeck = (deck: Deck, query: string) => {
  const needle = normalize(query)
  const fields = [
    deck.title,
    deck.commanders.join(" "),
    basename(deck.path).replace(/^[^_]+_/, ""),
    deck.relativePath,
  ].map(normalize)
  const exact = fields.some((field) => field === needle)
  const partial = fields.some((field) => field.includes(needle))
  return { exact, partial }
}

export const chooseDeck = async (decks: Deck[], query?: string) => {
  const matches = query
    ? decks.filter((deck) => matchesDeck(deck, query).partial)
    : decks
  const exact = query
    ? matches.filter((deck) => matchesDeck(deck, query).exact)
    : []

  if (exact.length === 1) return exact[0]
  if (matches.length === 1) return matches[0]
  if (matches.length === 0) throw new Error(`no resolved deck matches "${query}"`)
  if (!process.stdin.isTTY) {
    throw new Error(
      `${query ? `deck "${query}" is ambiguous` : "a deck selection is required"}:\n` +
        matches.map((deck, index) => `  ${index + 1}. ${deckLabel(deck)}`).join("\n"),
    )
  }

  const input = createInterface({ input: process.stdin, output: process.stderr })
  try {
    process.stderr.write(
      `\n${query ? `Choose a deck for "${query}"` : "Choose a deck"}:\n` +
        matches.map((deck, index) => `  ${index + 1}. ${deckLabel(deck)}`).join("\n") +
        "\n",
    )
    while (true) {
      const answer = await input.question(`Selection [1-${matches.length}]: `)
      const index = Number(answer) - 1
      if (Number.isInteger(index) && matches[index]) return matches[index]
      process.stderr.write("Enter one of the listed numbers.\n")
    }
  } finally {
    input.close()
  }
}

const parseArgs = (args: string[]): Options => {
  const options: Options = {
    apply: false,
    bottom: [],
    deckQueries: [],
    format: "markdown",
    list: false,
    mulligans: "0,0,0,0",
    repo: ROOT,
    seed: 1729,
    turns: 12,
  }

  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index]
    const value = () => {
      const next = args[index + 1]
      if (!next) throw new Error(`${arg} needs a value`)
      index += 1
      return next
    }
    if (arg === "--apply") options.apply = true
    else if (arg === "--bottom") options.bottom.push(value())
    else if (arg === "--format") options.format = value() as Options["format"]
    else if (arg === "--list") options.list = true
    else if (arg === "--mulligans") options.mulligans = value()
    else if (arg === "--out") options.out = value()
    else if (arg === "--repo") options.repo = resolve(value())
    else if (arg === "--seed") options.seed = Number(value())
    else if (arg === "--turns") options.turns = Number(value())
    else if (arg === "--help" || arg === "-h") {
      console.log(`Usage:
  bun run table:deal -- "Deck one" "Deck two" "Deck three" "Deck four" [options]
  bun run table:deal

Deck names may be brew titles, commanders, or folder slugs. Missing or
ambiguous names open a numbered selection when run interactively.

Options:
  --list                    List resolved decks
  --seed NUMBER             Shuffle seed (default: 1729)
  --turns NUMBER            Play through this turn (default: 12)
  --format markdown|json    Preview format (default: markdown)
  --apply                   Write opening game state
  --mulligans 0,1,0,0      Kept candidate per seat
  --bottom p2=Card,Card     London bottoms; repeat per seat
  --out PATH                Write output to a file
  --repo PATH               Repository root`)
      process.exit(0)
    } else if (arg.startsWith("-")) throw new Error(`unknown option: ${arg}`)
    else options.deckQueries.push(arg)
  }

  if (!Number.isInteger(options.seed) || options.seed < 0) {
    throw new Error("--seed must be a non-negative integer")
  }
  if (!["markdown", "json"].includes(options.format)) {
    throw new Error("--format must be markdown or json")
  }
  if (options.deckQueries.length > 4) throw new Error("pass at most four deck names")
  if (!Number.isInteger(options.turns) || options.turns < 1) {
    throw new Error("--turns must be a positive integer")
  }
  return options
}

export const seededRandom = (seed: number) => {
  let state = seed >>> 0
  return () => {
    state += 0x6d2b79f5
    let value = state
    value = Math.imul(value ^ (value >>> 15), value | 1)
    value ^= value + Math.imul(value ^ (value >>> 7), value | 61)
    return ((value ^ (value >>> 14)) >>> 0) / 4294967296
  }
}

export const shuffle = (cards: string[], random: () => number) => {
  const shuffled = [...cards]
  for (let index = shuffled.length - 1; index > 0; index -= 1) {
    const swap = Math.floor(random() * (index + 1))
    const card = shuffled[index]
    shuffled[index] = shuffled[swap]
    shuffled[swap] = card
  }
  return shuffled
}

const dealCandidates = (library: string[], random: () => number) =>
  Array.from({ length: MULLIGAN_MAX + 1 }, (_, mulligans) => {
    const shuffled = shuffle(library, random)
    return {
      mulligans,
      hand: shuffled.slice(0, 7),
      library: shuffled.slice(7),
    }
  })

const imageUris = (cache: Json) => {
  if (cache.image_uris?.small) return cache.image_uris
  return cache.card_faces?.find((face: Json) => face.image_uris?.small)?.image_uris ?? {}
}

const joinedFaces = (cache: Json, field: string) =>
  (cache.card_faces ?? []).map((face: Json) => face[field] || "").filter(Boolean).join(" // ")

const statLine = (cache: Json) => {
  if (cache.power != null) return `${cache.power}/${cache.toughness}`
  return (cache.card_faces ?? [])
    .filter((face: Json) => face.power != null)
    .map((face: Json) => `${face.power}/${face.toughness}`)
    .join(" // ")
}

const catalogEntry = async (entry: Json, repo: string) => {
  const card = entry.card ?? {}
  const cache = typeof entry.cache === "string"
    ? await readJson(join(repo, entry.cache)).catch(() => ({}))
    : {}
  const uris = imageUris(cache)
  return {
    scryfall_uri: cache.scryfall_uri || entry.scryfall_uri || "",
    image_small: uris.small || "",
    image_normal: uris.normal || uris.small || "",
    type_line: cache.type_line || card.type_line || joinedFaces(cache, "type_line"),
    mana_cost: cache.mana_cost || card.mana_cost || joinedFaces(cache, "mana_cost"),
    oracle_text: cache.oracle_text || card.oracle_text || joinedFaces(cache, "oracle_text"),
    stats: statLine(cache),
  }
}

const buildCatalog = async (seats: Seat[], repo: string) => {
  const entries = new Map<string, Json>()
  for (const seat of seats) {
    for (const entry of seat.entries) {
      if (typeof entry.name === "string" && !entries.has(entry.name)) {
        entries.set(entry.name, entry)
      }
    }
  }
  return Object.fromEntries(
    await Promise.all(
      [...entries].map(async ([name, entry]) => [name, await catalogEntry(entry, repo)]),
    ),
  )
}

const loadTokens = async (deckPath: string) => {
  try {
    const manifest = await readJson(join(deckPath, "tokens.json"))
    if (manifest.schema !== 1) throw new Error("unsupported token schema")
    return {
      tokenSources: manifest.produced_by ?? {},
      tokens: manifest.tokens ?? {},
    }
  } catch {
    return { tokenSources: {}, tokens: {} }
  }
}

const tableTokens = (seats: Seat[]) => ({
  token_sources: Object.fromEntries(
    seats.map((seat) => [seat.id, seat.tokenSources]),
  ),
  tokens: Object.assign({}, ...seats.map((seat) => seat.tokens)),
})

const loadSeats = async (decks: Deck[], seed: number) => {
  const random = seededRandom(seed)
  return Promise.all(
    decks.map(async (deck, index): Promise<Seat> => {
      const { library, entries } = await loadManifest(deck.path)
      const tokenData = await loadTokens(deck.path)
      return {
        ...deck,
        ...tokenData,
        id: SEAT_IDS[index],
        color: SEAT_COLORS[index],
        entries,
        candidates: dealCandidates(library, random),
      }
    }),
  )
}

const renderMarkdown = (seats: Seat[], seed: number) => {
  const lines = [
    `# Table deal — seed ${seed}`,
    "",
    `Turn order: ${seats.map((seat) => `${seat.id} ${seat.title}`).join(" → ")}`,
  ]
  for (const seat of seats) {
    lines.push("", `## ${seat.id} — ${seat.title}`, `Commander: ${seat.commanders.join(", ")}`)
    for (const candidate of seat.candidates) {
      const label = candidate.mulligans === 0 ? "keep seven" : `mulligan ${candidate.mulligans}`
      lines.push("", `### ${label}`, `- Hand: ${candidate.hand.join(", ")}`)
    }
  }
  return `${lines.join("\n")}\n`
}

const publicSeat = (seat: Seat, candidates = false) => ({
  id: seat.id,
  name: seat.title,
  deck: seat.relativePath,
  commanders: seat.commanders,
  plan: seat.plan,
  color: seat.color,
  ...(candidates
    ? {
        candidates: seat.candidates.map(({ mulligans, hand, library }) => ({
          mulligans,
          hand,
          library_count: library.length,
        })),
      }
    : {}),
})

const previewPayload = async (seats: Seat[], options: Options) => ({
  seed: options.seed,
  seats: seats.map((seat) => publicSeat(seat, true)),
  catalog: await buildCatalog(seats, options.repo),
  ...tableTokens(seats),
  _libraries: Object.fromEntries(
    seats.flatMap((seat) =>
      seat.candidates.map((candidate) => [
        `${seat.id}:${candidate.mulligans}`,
        { hand: candidate.hand, library: candidate.library },
      ]),
    ),
  ),
})

const parseBottoms = (values: string[]) => {
  const bottoms = Object.fromEntries(SEAT_IDS.map((seat) => [seat, [] as string[]]))
  for (const value of values) {
    const split = value.indexOf("=")
    if (split < 0) throw new Error(`bottom must be seat=Card,Card: ${value}`)
    const seat = value.slice(0, split).trim()
    if (!(seat in bottoms)) throw new Error(`unknown seat in --bottom: ${seat}`)
    bottoms[seat] = value
      .slice(split + 1)
      .split(",")
      .map((card) => card.trim())
      .filter(Boolean)
  }
  return bottoms
}

const applyLondon = (candidate: Candidate, bottom: string[]) => {
  if (bottom.length !== candidate.mulligans) {
    throw new Error(
      `mulligan ${candidate.mulligans} needs ${candidate.mulligans} bottom card(s)`,
    )
  }
  const hand = [...candidate.hand]
  const library = [...candidate.library]
  for (const card of bottom) {
    const index = hand.indexOf(card)
    if (index < 0) throw new Error(`cannot bottom "${card}"; hand is ${hand.join(", ")}`)
    hand.splice(index, 1)
    library.push(card)
  }
  return { hand, library }
}

const emptyPlayer = (
  commanders: string[],
  hand: string[],
  libraryCount: number,
  seatId: string,
) => ({
  life: 40,
  poison: 0,
  commander_damage: Object.fromEntries(
    SEAT_IDS.filter((seat) => seat !== seatId).map((seat) => [seat, 0]),
  ),
  commander_tax: 0,
  library_count: libraryCount,
  hand,
  battlefield: [],
  graveyard: [],
  exile: [],
  command: commanders,
  revealed_top: [],
})

const applyGame = async (seats: Seat[], options: Options) => {
  const counts = options.mulligans.split(",").map((part) => Number(part.trim()))
  if (
    counts.length !== 4 ||
    counts.some((count) => !Number.isInteger(count) || count < 0 || count > MULLIGAN_MAX)
  ) {
    throw new Error("--mulligans must be four integers from 0 to 2")
  }
  const bottoms = parseBottoms(options.bottom)
  const players: Json = {}
  const libraries: Json = {}
  const publicSeats: Json[] = []
  const keeps: Json[] = []

  seats.forEach((seat, index) => {
    const mulligans = counts[index]
    const candidate = seat.candidates.find((item) => item.mulligans === mulligans)!
    const kept = applyLondon(candidate, bottoms[seat.id])
    players[seat.id] = emptyPlayer(
      seat.commanders,
      kept.hand,
      kept.library.length,
      seat.id,
    )
    libraries[seat.id] = kept.library
    publicSeats.push({ ...publicSeat(seat), mulligans })
    keeps.push({
      turn: 0,
      phase: "setup",
      seat: seat.id,
      kind: mulligans === 0 ? "keep" : "mulligan",
      summary:
        mulligans === 0
          ? `${seat.title} keeps 7`
          : `${seat.title} mulligans ${mulligans}, bottoms ${bottoms[seat.id].join(", ")}`,
      cards: kept.hand,
      notes: "",
    })
  })

  const openingState = {
    active: "p1",
    turn: 0,
    phase: "setup",
    stack: [],
    players,
  }
  const events = [
    {
      turn: 0,
      phase: "setup",
      seat: null,
      kind: "setup",
      summary: "Four players keep. Turn-one draw is on.",
      cards: [],
      notes: "",
    },
    ...keeps,
  ].map((event, id) => ({
    id,
    ...event,
    state: structuredClone(openingState),
  }))

  return {
    schema: 1,
    seed: options.seed,
    starting_life: 40,
    headline: "In progress",
    horizon: { throughTurn: options.turns },
    result: {
      winner: null,
      ended: "truncated",
      turn: 0,
      summary: `Opening keeps only; play through turn ${options.turns} and overwrite this file.`,
    },
    seats: publicSeats,
    catalog: await buildCatalog(seats, options.repo),
    ...tableTokens(seats),
    events,
    _libraries: libraries,
  }
}

const main = async () => {
  const options = parseArgs(Bun.argv.slice(2))
  const available = await discoverDecks(options.repo)
  if (options.list) {
    console.log(available.map(deckLabel).join("\n"))
    return
  }

  const selected: Deck[] = []
  for (let index = 0; index < 4; index += 1) {
    const remaining = available.filter((deck) => !selected.includes(deck))
    selected.push(await chooseDeck(remaining, options.deckQueries[index]))
  }

  const seats = await loadSeats(selected, options.seed)
  let text: string
  if (options.apply) {
    text = JSON.stringify(await applyGame(seats, options), null, 2)
  } else if (options.format === "json") {
    text = JSON.stringify(await previewPayload(seats, options), null, 2)
  } else {
    text = renderMarkdown(seats, options.seed)
  }

  if (options.out) {
    const out = resolve(options.out)
    await mkdir(dirname(out), { recursive: true })
    await writeFile(out, text)
  } else {
    process.stdout.write(text.endsWith("\n") ? text : `${text}\n`)
  }
}

if (import.meta.main) {
  main().catch((error) => {
    console.error(`ERROR: ${error instanceof Error ? error.message : error}`)
    process.exitCode = 1
  })
}

File: .agents/skills/simulate-table/scripts/deal_table.py
#!/usr/bin/env python3
"""Deal a four-player Commander table from resolved decks."""

from __future__ import annotations

import argparse
import json
import random
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[4]
SEAT_IDS = ("p1", "p2", "p3", "p4")
SEAT_COLORS = ("#c45c26", "#2f6f64", "#4a5d9e", "#8a3d6b")
MULLIGAN_MAX = 2


def is_library_card(entry: dict) -> bool:
    return not any(
        category == "Commander{top}" or "{noDeck}" in category
        for category in entry.get("categories", [])
    )


def load_manifest(deck: Path) -> tuple[list[str], list[str], list[dict]]:
    manifest_path = deck if deck.name == "cards.json" else deck / "cards.json"
    if not manifest_path.is_file():
        raise ValueError(f"missing resolved manifest: {manifest_path}")

    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    library: list[str] = []
    commanders: list[str] = []
    entries: list[dict] = []

    for entry in manifest.get("cards", []):
        name = entry.get("name")
        quantity = entry.get("quantity")
        categories = entry.get("categories", [])
        if not isinstance(name, str) or not isinstance(quantity, int) or quantity < 1:
            raise ValueError(f"invalid manifest entry: {entry!r}")
        entries.append(entry)
        if "Commander{top}" in categories:
            commanders.extend([name] * quantity)
        elif is_library_card(entry):
            library.extend([name] * quantity)

    if len(library) not in (98, 99):
        raise ValueError(
            f"{deck}: expected a 99-card library or 98 with two commanders, "
            f"found {len(library)}"
        )
    if len(commanders) not in (1, 2):
        raise ValueError(f"{deck}: expected one or two commanders, found {len(commanders)}")
    if len(library) + len(commanders) != 100:
        raise ValueError(
            f"{deck}: expected 100 total cards, found {len(library) + len(commanders)}"
        )
    return library, commanders, entries


def deck_title(deck: Path) -> str:
    readme = deck / "README.md"
    if readme.is_file():
        first = readme.read_text(encoding="utf-8").splitlines()[0]
        match = re.search(r"—\s+(.+)$", first)
        if match:
            return match.group(1).strip()
    return deck.name


def primer_plan(deck: Path) -> str:
    readme = deck / "README.md"
    if not readme.is_file():
        return ""
    for line in readme.read_text(encoding="utf-8").splitlines():
        if line.startswith("**Primary plan:**"):
            return re.sub(r"\*\*|\[([^\]]+)\]\([^)]+\)", r"\1", line).replace(
                "Primary plan:", ""
            ).strip()
    return ""


def image_uris(cache: dict) -> dict:
    uris = cache.get("image_uris")
    if isinstance(uris, dict) and uris.get("small"):
        return uris
    for face in cache.get("card_faces") or []:
        face_uris = face.get("image_uris") if isinstance(face, dict) else None
        if isinstance(face_uris, dict) and face_uris.get("small"):
            return face_uris
    return {}


def joined_faces(cache: dict, field: str) -> str:
    """Split and modal cards carry text per face rather than at the top level."""
    parts = [
        face.get(field) or ""
        for face in cache.get("card_faces") or []
        if isinstance(face, dict)
    ]
    return " // ".join(part for part in parts if part)


def stat_line(cache: dict) -> str:
    if cache.get("power") is not None:
        return f"{cache['power']}/{cache['toughness']}"
    stats = [
        f"{face['power']}/{face['toughness']}"
        for face in cache.get("card_faces") or []
        if isinstance(face, dict) and face.get("power") is not None
    ]
    return " // ".join(stats)


def catalog_entry(entry: dict, repo: Path) -> dict:
    card = entry.get("card") or {}
    cache_rel = entry.get("cache")
    cache: dict = {}
    if isinstance(cache_rel, str):
        cache_path = repo / cache_rel
        if cache_path.is_file():
            cache = json.loads(cache_path.read_text(encoding="utf-8"))
    uris = image_uris(cache)
    return {
        "scryfall_uri": cache.get("scryfall_uri") or entry.get("scryfall_uri") or "",
        "image_small": uris.get("small") or "",
        "image_normal": uris.get("normal") or uris.get("small") or "",
        "type_line": (
            cache.get("type_line")
            or card.get("type_line")
            or joined_faces(cache, "type_line")
        ),
        "mana_cost": (
            cache.get("mana_cost")
            or card.get("mana_cost")
            or joined_faces(cache, "mana_cost")
        ),
        "oracle_text": (
            cache.get("oracle_text")
            or card.get("oracle_text")
            or joined_faces(cache, "oracle_text")
        ),
        "stats": stat_line(cache),
    }


def build_catalog(entries_by_seat: list[list[dict]], repo: Path) -> dict[str, dict]:
    catalog: dict[str, dict] = {}
    for entries in entries_by_seat:
        for entry in entries:
            name = entry.get("name")
            if isinstance(name, str) and name not in catalog:
                catalog[name] = catalog_entry(entry, repo)
    return catalog


def deal_candidates(
    library: list[str],
    rng: random.Random,
    *,
    mulligans: int = MULLIGAN_MAX,
) -> list[dict]:
    candidates = []
    for mulligan_count in range(mulligans + 1):
        shuffled = library.copy()
        rng.shuffle(shuffled)
        hand = shuffled[:7]
        rest = shuffled[7:]
        candidates.append({
            "mulligans": mulligan_count,
            "hand": hand,
            "library": rest,
        })
    return candidates


def parse_bottom(values: list[str]) -> dict[str, list[str]]:
    bottoms: dict[str, list[str]] = {seat: [] for seat in SEAT_IDS}
    for raw in values:
        if "=" not in raw:
            raise ValueError(f"bottom must be seat=Card,Card: {raw!r}")
        seat, cards = raw.split("=", 1)
        seat = seat.strip()
        if seat not in bottoms:
            raise ValueError(f"unknown seat in --bottom: {seat}")
        names = [name.strip() for name in cards.split(",") if name.strip()]
        bottoms[seat] = names
    return bottoms


def apply_london(candidate: dict, bottom: list[str]) -> tuple[list[str], list[str]]:
    hand = list(candidate["hand"])
    library = list(candidate["library"])
    expected = candidate["mulligans"]
    if len(bottom) != expected:
        raise ValueError(
            f"mulligan {expected} needs {expected} bottom card(s), got {bottom!r}"
        )
    remaining = hand
    for name in bottom:
        if name not in remaining:
            raise ValueError(f"cannot bottom {name!r}; hand is {hand}")
        remaining.remove(name)
        library.append(name)
    return remaining, library


def empty_player(commanders: list[str], hand: list[str], library_count: int, seat_id: str) -> dict:
    return {
        "life": 40,
        "poison": 0,
        "commander_damage": {seat: 0 for seat in SEAT_IDS if seat != seat_id},
        "commander_tax": 0,
        "library_count": library_count,
        "hand": hand,
        "battlefield": [],
        "graveyard": [],
        "exile": [],
        "command": commanders,
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Deal a four-player Commander table from cards.json manifests"
    )
    parser.add_argument(
        "decks",
        nargs=4,
        type=Path,
        help="four deck directories in turn order",
    )
    parser.add_argument("--seed", type=int, default=1729)
    parser.add_argument("--format", choices=("markdown", "json"), default="markdown")
    parser.add_argument("--apply", action="store_true")
    parser.add_argument("--mulligans", default="0,0,0,0")
    parser.add_argument(
        "--bottom",
        action="append",
        default=[],
        help="London bottoms, e.g. p2=Mountain,Forest (repeatable)",
    )
    parser.add_argument("--out", type=Path)
    parser.add_argument("--repo", type=Path, default=ROOT)
    args = parser.parse_args()
    if args.seed < 0:
        parser.error("--seed cannot be negative")
    return args


def load_seats(decks: list[Path], seed: int) -> list[dict]:
    rng = random.Random(seed)
    seats = []
    for index, deck in enumerate(decks):
        deck = deck.resolve()
        library, commanders, entries = load_manifest(deck)
        candidates = deal_candidates(library, rng)
        seats.append({
            "id": SEAT_IDS[index],
            "name": deck_title(deck),
            "deck": str(deck.relative_to(ROOT) if ROOT in deck.parents else deck),
            "commanders": commanders,
            "plan": primer_plan(deck),
            "color": SEAT_COLORS[index],
            "entries": entries,
            "candidates": candidates,
        })
    return seats


def render_preview_markdown(seats: list[dict], seed: int) -> str:
    lines = [
        f"# Table deal — seed {seed}",
        "",
        "Turn order: " + " → ".join(f"{seat['id']} {seat['name']}" for seat in seats),
    ]
    for seat in seats:
        lines.extend(["", f"## {seat['id']} — {seat['name']}"])
        lines.append(f"Commander: {', '.join(seat['commanders'])}")
        for candidate in seat["candidates"]:
            label = "keep seven" if candidate["mulligans"] == 0 else (
                f"mulligan {candidate['mulligans']}"
            )
            lines.extend([
                "",
                f"### {label}",
                f"- Hand: {', '.join(candidate['hand'])}",
            ])
    return "\n".join(lines) + "\n"


def preview_payload(seats: list[dict], seed: int, repo: Path) -> dict:
    catalog = build_catalog([seat["entries"] for seat in seats], repo)
    return {
        "seed": seed,
        "seats": [
            {
                "id": seat["id"],
                "name": seat["name"],
                "deck": seat["deck"],
                "commanders": seat["commanders"],
                "plan": seat["plan"],
                "color": seat["color"],
                "candidates": [
                    {
                        "mulligans": candidate["mulligans"],
                        "hand": candidate["hand"],
                        "library_count": len(candidate["library"]),
                    }
                    for candidate in seat["candidates"]
                ],
            }
            for seat in seats
        ],
        "catalog": catalog,
        "_libraries": {
            f"{seat['id']}:{candidate['mulligans']}": {
                "hand": candidate["hand"],
                "library": candidate["library"],
            }
            for seat in seats
            for candidate in seat["candidates"]
        },
    }


def apply_game(seats: list[dict], args: argparse.Namespace) -> dict:
    counts = [int(part.strip()) for part in args.mulligans.split(",")]
    if len(counts) != 4 or any(count < 0 or count > MULLIGAN_MAX for count in counts):
        raise ValueError("--mulligans must be four integers from 0 to 2")
    bottoms = parse_bottom(args.bottom)
    catalog = build_catalog([seat["entries"] for seat in seats], args.repo)
    public_seats = []
    players = {}
    events = []
    libraries: dict[str, list[str]] = {}

    for seat, mulligans in zip(seats, counts):
        candidate = next(
            item for item in seat["candidates"] if item["mulligans"] == mulligans
        )
        hand, library = apply_london(candidate, bottoms.get(seat["id"], []))
        libraries[seat["id"]] = library
        public_seats.append({
            "id": seat["id"],
            "name": seat["name"],
            "deck": seat["deck"],
            "commanders": seat["commanders"],
            "plan": seat["plan"],
            "mulligans": mulligans,
            "color": seat["color"],
        })
        players[seat["id"]] = empty_player(seat["commanders"], hand, len(library), seat["id"])
        summary = (
            f"{seat['name']} keeps 7"
            if mulligans == 0
            else f"{seat['name']} mulligans {mulligans}, bottoms {', '.join(bottoms[seat['id']])}"
        )
        events.append({
            "id": len(events),
            "turn": 0,
            "phase": "setup",
            "seat": seat["id"],
            "kind": "keep" if mulligans == 0 else "mulligan",
            "summary": summary,
            "cards": hand,
            "notes": "",
            "state": None,
        })

    opening_state = {
        "active": "p1",
        "turn": 0,
        "phase": "setup",
        "stack": [],
        "players": players,
    }
    events.insert(0, {
        "id": 0,
        "turn": 0,
        "phase": "setup",
        "seat": None,
        "kind": "setup",
        "summary": "Four players keep. Turn-one draw is on.",
        "cards": [],
        "notes": "",
        "state": opening_state,
    })
    for index, event in enumerate(events[1:], start=1):
        event["id"] = index
        event["state"] = {
            "active": "p1",
            "turn": 0,
            "phase": "setup",
            "stack": [],
            "players": json.loads(json.dumps(players)),
        }

    return {
        "schema": 1,
        "seed": args.seed,
        "starting_life": 40,
        "headline": "In progress",
        "result": {
            "winner": None,
            "ended": "truncated",
            "turn": 0,
            "summary": "Opening keeps only; play the game and overwrite this file.",
        },
        "seats": public_seats,
        "catalog": catalog,
        "events": events,
        "_libraries": libraries,
    }


def main() -> int:
    args = parse_args()
    try:
        seats = load_seats(args.decks, args.seed)
        if args.apply:
            payload = apply_game(seats, args)
            text = json.dumps(payload, indent=2)
        elif args.format == "json":
            payload = preview_payload(seats, args.seed, args.repo)
            text = json.dumps(payload, indent=2)
        else:
            text = render_preview_markdown(seats, args.seed)
    except (OSError, ValueError, json.JSONDecodeError) as error:
        print(f"ERROR: {error}", file=sys.stderr)
        return 1

    if args.out:
        args.out.parent.mkdir(parents=True, exist_ok=True)
        args.out.write_text(text, encoding="utf-8")
    else:
        print(text, end="" if text.endswith("\n") else "\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
